File: TUnit.TestProject/EnumerableDataSourceDrivenTests.cs
using TUnit.TestProject.Attributes;

namespace TUnit.TestProject;

[EngineTest(ExpectedResult.Pass)]
public class EnumerableDataSourceDrivenTests
{
    [Test]
    [MethodDataSource(nameof(SomeMethod))]
    public async Task DataSource_Method(int value)
    {
        await Assert.That(value).IsBetween(1, 5).WithInclusiveBounds();
    }

    [Test]
    [MethodDataSource(nameof(SomeMethod))]
    public async Task DataSource_Method2(int value)
    {
        await Assert.That(value).IsBetween(1, 5).WithInclusiveBounds();
    }

    [Test]
    [MethodDataSource(nameof(MethodWithBaseReturn))]
    public void DataSource_WithBaseReturn(BaseValue value)
    {
    }

    public static IEnumerable<int> SomeMethod() => [1, 2, 3, 4, 5];

    public static List<Func<BaseValue>> MethodWithBaseReturn() =>
    [
        () => new ConcreteValue(),
        () => new ConcreteValue2()
    ];

    public abstract class BaseValue;

    public class ConcreteValue : BaseValue;
    public class ConcreteValue2 : BaseValue;
}

File: TUnit.TestProject.FSharp/EnumerableDataSourceDrivenTests.fs
namespace TUnit.TestProject.FSharp

open System.Collections.Generic
open System.Threading.Tasks
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Core

// F# equivalent of EnumerableDataSourceDrivenTests.cs

type BaseValue() = class end

type ConcreteValue() =
    inherit BaseValue()

type ConcreteValue2() =
    inherit BaseValue()

type EnumerableDataSourceDrivenTests() =
    [<Test>]
    [<MethodDataSource("SomeMethod")>]
    member _.DataSource_Method(value: int) : Task =
        Assert.That(value).IsEqualTo(1)

    [<Test>]
    [<MethodDataSource("SomeMethod")>]
    member _.DataSource_Method2(value: int) : Task =
        Assert.That(value).IsEqualTo(1)

    [<Test>]
    [<MethodDataSource("MethodWithBaseReturn")>]
    member _.DataSource_WithBaseReturn(value: BaseValue) =
        ()

    static member SomeMethod() = seq { 1; 2; 3; 4; 5 }

    static member MethodWithBaseReturn() =
        [ fun () -> ConcreteValue() :> BaseValue
          fun () -> ConcreteValue2() :> BaseValue ]
